### PR TITLE
Command Line Client

### PR DESCRIPTION
**Pull request by samuel-tate**
## Related Issues
#22 

## Testing locally

- set up and activate virtual environment

```
$ virtualenv venv
$ . venv/bin/activate
```

- install local pyrefinebio package as a pip package

```
$ pip install --editable .
```

- use the cli~

```
$ refinebio download-dataset --path "./test.zip" --email-address "foo@bar.net" --experiments "GSE74410" 
```

## Example runs

Right now there aren't really good outputs for the download commands.
Not sure if I should prioritize outputs for those functions as a part of this PR.

<details>
<summary>help example</summary>
<p>

```
(venv) ➜  refinebio-py git:(state/cli) refinebio help "Sample.search"                            
Retrieve a list of Samples based on various filters

        Returns:
            list of Sample

        Parameters:
            ordering (str): which field to use when ordering the results

            title (str): filter based on the Sample's title

            organism (str): filter based on the Organism that the Sample was taken from

            source_database (str): filter based on the publically available repository
                                   that the Sample was taken from

            source_archive_url (str): filter based on Sample's source url

            has_raw (bool): filter based on if the Sample had raw data available in the source database

            platform_name (str): filter based on the name of the Platform that was used to assay the Sample

            technology (str): filter based on the technology that was used to assay the Sample

            manufacturer (str): filter based on the manufacturer of the technology that was used to assay the Sample

            sex (str): filter based on the sex information provided by the submitter

            age (number): filter based on the age informtaion provided by the submitter

            specimen_part (str): filter based on the part of the specimen reported by the submitter

            genotype (str): filter based on the genotype of the subject that the Sample was taken from

            disease (str): filter based on the disease information provided by the submitter

            disease_stage (str): filter based on the disease stage information provided by the submitter

            cell_line (str): filter based on the cell line used for the sample

            treatment (str): filter based on the treatment information provided by the submitter

            race (str): filter based on the race information provided by the submitter

            subject (str): filter based on the subject identifier information provided by the submitter

            compound (str): filter based on the compound information provided by the submitter

            time (str): filter based on the time information provided by the submitter

            is_processed (bool): filter based on if the Sample has been processed

            is_public (bool): filter based on if the Sample is public

            limit (int): number of results to return per page.

            offset (int): the initial index from which to return the results.

            dataset_id (str): filter based on the Dataset ids that the Sample has been added to

            experiment_accession_code (str): filter based on the Dxperiments that are associated
                                             with the Sample

            accession_codes (str): filter based on multiple accession codes at once
```

</p>
</details>

### CODE REVIEW

File: pyrefinebio/exceptions.py
Comment (kurtwheeler) [inline pyrefinebio/exceptions.py:4]: What's the reason for this change and what's it doing?

File: pyrefinebio/script.py
Comment (kurtwheeler) [inline pyrefinebio/script.py:53]: Any chance we can make this an unnamed arg? `refinebio help "Sample.search"` seems easier to type.

File: pyrefinebio/script.py
Comment (kurtwheeler) [inline pyrefinebio/script.py:65]: These lines are too long, they're kind of hard to read. [PEP8](https://www.python.org/dev/peps/pep-0008/) says:

```
The default wrapping in most tools disrupts the visual structure of the code, making it more difficult to understand. The limits are chosen to avoid wrapping in editors with the window width set to 80, even if the tool places a marker glyph in the final column when wrapping lines. Some web based tools may not offer dynamic line wrapping at all.

Some teams strongly prefer a longer line length. For code maintained exclusively or primarily by a team that can reach agreement on this issue, it is okay to increase the line length limit up to 99 characters, provided that comments and docstrings are still wrapped at 72 characters.
```

I fall into the "prefer a longer line length" of 100. If reformatting these to be shorter seems tedious, you could just run [black](https://github.com/psf/black) on them with `black --line-length=100 .`

File: pyrefinebio/script.py
Comment (kurtwheeler) [inline pyrefinebio/script.py:61]: I think this should specify that the data should be in JSON format.

File: pyrefinebio/script.py
Comment (kurtwheeler) [inline pyrefinebio/script.py:95]:
```suggestion
@click.option("--organism", help="The name of the Organism for the Compendium you want to download")
```

File: pyrefinebio/script.py
Comment (kurtwheeler) [inline pyrefinebio/script.py:94]: Could this default to the current directory? I guess it'd need to wait on https://github.com/AlexsLemonade/refinebio-py/issues/31

Comment (kurtwheeler) [reply pyrefinebio/script.py:94]: I added a comment there.

File: pyrefinebio/script.py
Comment (kurtwheeler) [inline pyrefinebio/script.py:117]:
```suggestion
@click.option("--organism", help="The name of the Organism for the Compendium you want to download")
```

File: pyrefinebio/script.py
Comment (kurtwheeler) [inline pyrefinebio/script.py:121]: Maybe we can link to http://docs.refine.bio/en/latest/main_text.html#rna-seq-sample-compendia to provide more context?

File: pyrefinebio/script.py
Comment (kurtwheeler) [inline pyrefinebio/script.py:103]: Might be nice to link to http://docs.refine.bio/en/latest/main_text.html#normalized-compendia

**Review by kurtwheeler — CHANGES_REQUESTED** [review]
This is looking good so far! I'd like to know how testing it from the CLI went.

I've also thought more about the token situation. I think having a higher level function for them is still a good idea, but we can also totally abstract the token away. I think that on any function that requires a token, if there's not a saved token then we should just prompt the user to agree to our ToC (providing a link), then once they type in yes we just create a token for them, we can offer an option to save this so they don't have to do so every time.

File: pyrefinebio/exceptions.py
Comment (samuel-tate) [reply pyrefinebio/exceptions.py:4]: the goal was to expose the fully formatted message to be used like this:

```
try:
    something.something()
except pyrefinebio.exceptions.DownloadError as e:
    raise OtherException(e.message)
```

previously I would use `e.base_message` which would be weird unformatted stuff for some of the errors

I wasn't sure if this was the best way to do this, but it was the easiest for me for right now.
I'm thinking of refactoring all of these exceptions as a part of #26 anyway but we'll see.
if there is a better way lmk though!

File: pyrefinebio/script.py
Comment (kurtwheeler) [reply pyrefinebio/script.py:122]: I didn't mean to wait to re-review this long, but since we have a PR open for #31 maybe we can just wait until it lands to merge this so we can make this change here?

Comment (samuel-tate) [reply pyrefinebio/script.py]: added default path!

File: pyrefinebio/script.py
Comment (kurtwheeler) [inline pyrefinebio/script.py:57]: Does this get set as the helpstring by default?

**Comment by kurtwheeler** [comment]
Looks like the tests are expecting an error message that's a bit off.

File: pyrefinebio/script.py
Comment (samuel-tate) [reply pyrefinebio/script.py:57]: I just added `no_args_is_help=True` to this click command

now when you specify no args to `refinebio help` you get this:

image

lmk if this looks good to you

Comment (kurtwheeler) [reply pyrefinebio/script.py:57]: Ok sweet, this does look pretty good. However if I'm using this from the command line I'd actually probably want to know what my choices of sub-commands are, then I might see how to call the one I'm interested in by calling `refinebio help <MYCOMMAND>`.

Comment (samuel-tate) [reply pyrefinebio/script.py:57]: ohh ok I think you get that by doing `$ refinebio --help` or just `$ refinebio`

image

Then you could do `$ refinebio COMMAND --help` to get help for that command.

This help command really only gets help for the actual classes and functions in pyrefinebio. If you wanted to use it to get help for the cli commands you would have to do `$ refinebio help "script.dowload_dataset"` and in that case it only prints the docstring.

image

image

Comment (kurtwheeler) [reply pyrefinebio/script.py:57]: Ah ok, that makes sense then. Is there any way to have `refinebio help` do the same thing as `refinebio --help`? If I type the first one and it outputs, I might not think to try the second.

**Review by kurtwheeler — APPROVED** [review]
Sweet, this LGTM!

**Pull request merged**
